### ISSUE TITLE
Regridder to conserve horizontal mass fluxes

### ISSUE THREAD

**Issue by tclune** [GEOS-ESM/MAPL#637]
@sdeastham has requested a MAPL regridding option to provide exact regridding for horizontal mass fluxes  to a coarse grid from a fine grid that nests "exactly".   E.g.  C90 and C360.   The basic ideas is that conservation is then a simple process of summing the fluxes from the finer grid.   Generalization would be nice, but is not essential.

**Comment by tclune** [comment]
NOTE:   I encountered a surprise that will require more thought on our end when I started sketching out the process.   You are welcome to read below, as it is at least indicative of what needs to be done.       If I knew for certain how to ensure that ESMF conservative regridding would include the cells with the correct edges then this could just be a matter of setting up some custom weights  in an initialize method.   But ESMF works with overlapping areas which makes edges potentially fragile constructs under roundoff.   Our older system would work well for this, but ... it is the older system and not used.
----------------------

MAPL already has hooks to introduce variant regridders.     The capability is almost pointless now given that ESMF regridding is used in almost all cases, but fortunately it was not eliminated and remains as a means to implement the desired functionality.    The no longer used lat-lon to lat-lon regridder at 

https://github.com/GEOS-ESM/MAPL/blob/main/base/MAPL_LatLonToLatLonRegridder.F90

can be used as a rough guide.

Basically a new subclass (type extension) of AbstractRegridder must be implemented that provides concrete methods for  the various `regrid_*()` type-bound procedures and the `initialize()` method.    And actually, the scalar regridding can be ignored, as that case does not make sense.      I think we should also assume that the decomposition is "friendly" in that the edges to be summed together should all fall on a single process.  (Bonus points to at least add a check to enforce this.)

The first step would be to determine during initialize the nesting factor  N between the 2 grids.   Lots of heuristics suggest themselves.

Then in the vector regrid procedures, one would have outer loop over the coarse (target) grid points, and an inner loop over the contained segments of the fine (source) grid:
```f90
do k = 1, LM
     do j = 1, JM
          jj = 1 + (j-1)*N
          do i  = 1, IM
               m_y = 0
               do ii = 1 + (i-1)*N, i*N
                    m_y = m_ y + v_in(ii,j,k)
               end do
              v_out(i,j,k) = m_y
         end do
    end do
end do
```
There would be a similar loop for the x-fluxes which would have an inner loop over the j coordinate:
```f90
do k = 1, LM
     do i  = 1, IM
          ii = 1 + (i-1)*N
         do j = 1, JM
               m_x = 0
               do jj = 1 + (j-1)*N, j*N
                    m_x = m_ x + u_in(ii,j,k)
               end do
              u_out(i,j,k) = m_x
         end do
    end do
end do
```

The final step is to register the new subclass and a unique integer "method".   But that ties into the big caveat at the beginning of this post.

**Comment by sdeastham** [comment]
Thanks Tom! I want to make sure I understand correctly before I act. Am I right in thinking that this could currently be implemented using the "old" techniques without necessarily causing a problem, but that it will be non-trivial to develop an "ESMF-based" method? Or is there a way already in which the ESMF regridding will interfere?

**Comment by tclune** [comment]
It could have been implement in the older mechanism in a straightforward manner, but ... the older mechanism is not used in ExtData so it won't help us now.      Doing this as a custom ESMF regridder might be possible, but I currently don't know how to force ESMF to choose the right cells.     Setting the weights is otherwise straightforward, and we'd be able to loosen the constraint that both grids divide evenly into the process topology.

The way forward is probably to slightly extend the new manager mechanism to allow alternative regridders to be injected.    It has the ability to use them, just no interface to add them in in the first place.   I think what I want to do is change the RegriddingMethod into a derived type (currently it is an integer).   I can then make the existing support use EsmfRegriddingMethod, and introduce a new derived type for your case.   This lets us keep the single interface for adding regridders.    ExtData would then just need a modify the logic for how it specifies the RegriddingMethod argument.   Unfortunately, History would also need corresponding modifications.    Not terribly hard, but would need to be tested thoroughly before we add your case in.

It won't hurt for you to also request this capability from ESMF itself in the mean time.    On their end they probably have to add 2 different regridding methods:  one for Mx and one for My, as they don't recognize vectors.   I'm sure they cannot get it to you quickly, but it will allow MAPL to drop the custom code if/when they do add that capability.

**Comment by tclune** [comment]
Hmm. Had not realized that the line of dashes was making my "note" above bold and large font.  Had not meant to "scream".  Sorry about that.

Gave this a bit more thought and came back to the same conclusion.  Here are the implementation steps (and rough guidance on GMAO vs GCHP regarding who should implement):

1. (GMAO) Introduce new base class for regridding methods.       Possibly this can be the existing ESMF derived type, but often their use of "sequence" precludes type extension in the way that I envision.   The goal is to end up with 2 derived types that both extend some base type, with the ESMF regridding methods associated with one class and our extensions associated with the other.
2. (GMAO) Modify existing references to regridding methods to use the type from (1) associated with ESMF methods.
3. (GCHP) in ExtData where the regridding method is determined, detect if "horizontal conserve" (or whatever we call it) and encapsulate in the new derived type.
4. (GMAO) modify logic in new RegridderManager to check the type of the regridding method and either use `EsmfRegridder` (as it is now) or `CustomHorzConservativeRegridder` (or whatever we call it).
5. (GCHP) Implement a subclass of AbstractRegridder for this scenario along the lines described at the beginning of this thread.


Item (4) should eventually be improved so that we register prototypes.   But until there is more than one custom case this is not necessary.

I'll try to tackle at least (1) today.

**Comment by tclune** [comment]
Just a note to myself:  confirming that one cannot "extend" ESMF types:
```
/gpfsm/dswdev/tclune/GEOS-ESM/GEOSgcm/src/Shared/@MAPL/base/CustomRegridMethod.F90(8): error #8204: The parent of an extended type cannot have the BIND attribute or SEQUE\
NCE property.
   type, extends(ESMF_RegridMethod_Flag) :: CustomRegridMethod
-----------------^
```

**Comment by sdeastham** [comment]
That breakdown of responsibilities for implementation makes sense to me. Assuming I understand correctly, I'll wait on completion of (2) (or at least (1)) before embarking on (3) or (5).

**Comment by tclune** [comment]
I suspect I'll do most of everything in the end anyway - just to have something to exercise the new machinery.    But I'll leave it to you to test with real use cases.

Below is mostly catharsis for myself, but others may find the reasoning useful:

I did a fair bit on this on Monday, and decided I was making a mess by cleaning  and extending at the same time.   Started over Tuesday and did a quick sweep of cleaning some things that were simply in the way.  (Mostly cases where we were a bit lazy and had multiple modules and/or types declared in the same file.)     

I'm now fence sitting between two implementation choices.    In the older grid manager,  there was a look up table for the class of regridder to be used.   The key to the table was a triple  (grid type in, grid type out, and regrid method).   This reflected the granularity of the home-grown regridding methods in MAPL.     As we switched to ESMF, we could use the same class regardless of grid types, and eventually (albeit temporarily given the case at hand) for all regrid methods.    So the registry was removed.     I'm torn between re-introducing the registry (which might allow us to have a single RegridderManager again!), and converting the method specification from integer to derived types which would then be used to control the class of Regridder.

The latter probably involves fewer changes and meets the existing case reasonably well.   It does not quite work with arbitrary grid types though, which suggests the registry would be a more robust solution.       Alternatively, the old registry needs to get a bit smarter so that we don't have to explicitly list every possible combination of grid types and regrid methods that are supported by ESMF.   (If we have just lat-lon, CS, and tripolar, there are already O(50) combinations.)    

Possibly the right thing to do is to introduce a new method for the RegridderClass family which checks whether it can handle particular grid types and methods.     The EsmfRegridder can then basically ignore the grid type and just check that the method is one of the ones defined by ESMF.   The old home-grown regridders (still used in some bits of GEOS), can do a more explicit check. E.g., the LatLonToLatLonRegridder would check that both grids are, er, lat-lon, and that the method is either bilinear or conserve.

**Comment by tclune** [comment]
OK - the basic infrastructure is now in place on branch `feature/tclune/#637-horz-conserve-regridder`

There is a stub implementation of HorzFluxRegridder.F90 in MAPL/base.    I suspect I need to do the rest of the implementation, but it really should be very much like I sketched out at the beginning of this issue.

**Comment by tclune** [comment]
The intermediate progress is now merged into the develop branch, so no need to look at the feature branch.

**Comment by tclune** [comment]
Wow - it has already been 2 months since this was (re) opened!    I think I now finally have some time to address the issue.   I'll catch up on the weekly telecon to be sure I capture any work @sdeastham has done in the mean time.

**Comment by tclune** [comment]
This is nominally completed no on the feature/tclune/#637... branch.     Almost worked right out of the box - just needed some little tweaks for the declaration of ESMF arguments.    Tested with synthetic case using @bena-nasa 's ExtDataDriver.   Now could use some more thorough testing before being merged.

**Comment by sdeastham** [comment]
Great! Attempting a merge into GCHP now - I'll try to test this out with some mass fluxes over the next few days.

**Comment by sdeastham** [comment]
@tclune - merge went smoothly, and just ran the first test (regridding from C90 to C30). Results not quite as expected, so wanted to check that I'm doing things correctly:

1. Getting MFXIN and MFYIN through ExtData.rc with the following line:

`MFXIN;MFYIN 'kg s-1' N H F0;003000 none 1.00 MFX;MFY  ./Met_c90/tavg1_3d_asm_CSv-/%y4%m2/Heracles-5_4_p3.tavg1_3d_asm_CSv-.%y4%m2%d2_%h230z.nc4 2000-01-01T00:00:00P01:00`

2. Specifying MFXIN and MFYIN imports in our gridded component with (does this perhaps need a `rotation` keyword?):

```
      call MAPL_AddImportSpec ( gc,                                  &
           SHORT_NAME = 'MFXIN',                                     &
           LONG_NAME  = 'x-direction_mass_flux',                     &
           UNITS      = 'kg s-1',                                    &
           DIMS       = MAPL_DimsHorzVert,                           &
           VLOCATION  = MAPL_VLocationCenter,           RC=STATUS  )
      _VERIFY(STATUS)
```

At the moment, I get the result below. Top row is the true mass fluxes at c90, in the X (left) and Y (right) directions. Below is the result at c30; the increase in magnitude looks about right (~3x, although I truncated the color bars) but otherwise looks like perhaps there is an issue with rotation and/or placement?

![image](https://user-images.githubusercontent.com/6618732/107824746-fbd7cd00-6d4f-11eb-9ca3-9da69c7746a2.png)

**Comment by sdeastham** [comment]
Follow-up: this issue appears to be sensitive to the number of cores (see figure below, which shows a c90 simulation and then a c30 simulation using online regridding with 24, 12, and 6 cores). Most interestingly, the X and Y mass fluxes respond differently?

![image](https://user-images.githubusercontent.com/6618732/107835082-45cbad80-6d66-11eb-91f6-a66f4209c94c.png)

**Comment by tclune** [comment]
Sorry - I missed this email is a blizzard of other GEOS-ESM traffic.       I think you are correct that this is probably related to the "rotate" option.       Worst case is that I need to add a mechanism to deactivate some otherwise automatic treatment of vectors that operates at a higher level in the call sequence.

**Comment by tclune** [comment]
I spent a bit of time on this, but am still none the wiser.    The synthetic data I originally tested with was constant which clearly would not reveal any pattern like what you are seeing above.   But it does indicate that X and Y are not being mixed, which eliminates certain types of issues.

I'm going to try to modify the synthetic case to have more useful results.   I'll also speak with @bena-nasa to see if there is something outside regridding for vector cases that I've not properly understood.

**Comment by bena-nasa** [comment]
This made me think I should add something to my tester so that I could fill in something other than constant values so I added the ability to use the grid lats and lons from the grid in the expression used to fill in the variables when generating the synthetic data. I then tested this on top of Toms flux regridding branch and ran the test case I provided him utilizing the new flux regridding option. Instead of constants I put the lons in the x flux and the lats in the y flux. When I did this I found I was getting different results on 6 or 12 processors so something seems wrong.

**Comment by tclune** [comment]
OK - I'll double check my code again.    The decomposition dependence is an important clue.

In theory we should be able to look at 3 neighboring cells in the input data (with the right modulo wrt the nesting ratio) and add the fluxes by hand.   And then go see what is in the output grid at the corresponding position.

**Comment by bena-nasa** [comment]
Shouldn't this be ii, and not i on the right hand side of that expression:
https://github.com/GEOS-ESM/MAPL/blob/feature/tclune/%23637-horz-mass-flux-regrid/base/HorizontalFluxRegridder.F90#L144
Likewise for the y fluxes, j->jj?

**Comment by bena-nasa** [comment]
Yep, that fixed in my layout test. I get the same answer now on difference core counts.

**Comment by bena-nasa** [comment]
I pushed to the branch with this fix Seb if you want to pull and try again.

**Comment by sdeastham** [comment]
Looks good! Preliminary tests show that the online regridding

 a) reproduces the expected fluxes based on hand regridding;
 b) yields the same results with 6 and 24 cores;
 c) produces reasonable transport; and
 d) seems to result in pressure errors which are ~30 times smaller (at c30) than when using winds (compared to ~10 times smaller at c90).
 
 That last line is subject to (a lot) more testing, but seems encouraging!
